### PR TITLE
S3 uploader debugging

### PR DESCRIPTION
**Pull request by matthewfala**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -66,21 +66,40 @@ linux-plugins:
     	--CLOUDWATCH_PLUGIN_CLONE_URL=${CLOUDWATCH_PLUGIN_CLONE_URL} \
     	--CLOUDWATCH_PLUGIN_TAG=${CLOUDWATCH_PLUGIN_TAG} \
     	--CLOUDWATCH_PLUGIN_BRANCH=${CLOUDWATCH_PLUGIN_BRANCH} \
-		--DOCKER_BUILD_FLAGS=${DOCKER_BUILD_FLAGS}
+    	--DOCKER_BUILD_FLAGS=${DOCKER_BUILD_FLAGS}
 
+# Debug and debug init images
 .PHONY: main-debug
-main-debug: main-debug-base
+main-debug: debug-s3
```
Comment (PettitWesley) [inline Makefile:73]: So main-debug now is just an alias for main-debug-s3 and then it re-tags?

So there's no one target for all debug?

File: scripts/core_uploader.sh [hunk 2]
```diff
@@ -71,7 +69,6 @@ else
 
     # Zip the crashfiles
     zip -r $CRASH_FOLDER/${CORE_FILENAME}.all.zip $CRASH_FOLDER
-    mv /${CORE_FILENAME}.all.zip $CRASH_FOLDER/${CORE_FILENAME}.all.zip
```
Comment (PettitWesley) [inline scripts/core_uploader.sh:74]: why is this removed?

File: scripts/core_uploader.sh [hunk 3]
```diff
@@ -60,7 +59,6 @@ else
     #   - CORE_FILENAME.core
 
     # Process corefile, get stack trace and fluent bit executable. Add to zip.
-    mv `ls $CRASH_FOLDER/core*` ${CORE_FILENAME}.core
```
Comment (PettitWesley) [inline scripts/core_uploader.sh:63]: why?

File: scripts/dockerfiles/Dockerfile.init-debug-s3 [hunk 4]
```diff
@@ -41,7 +41,7 @@ CMD echo "AWS for Fluent Bit Container Image Version `cat /AWS_FOR_FLUENT_BIT_VE
     fi; \
     /init/fluent_bit_init_process; \
     chmod +x /init/invoke_fluent_bit.sh; \
-    export RANDOM_ID_VALUE=$(($RANDOM%99999))$(($RANDOM%99999))$(($RANDOM%99999)); \
-    echo "RANDOM_ID is set to $RANDOM_ID_VALUE"; \
-    env RANDOM_ID=$RANDOM_ID_VALUE /init/invoke_fluent_bit.sh; \
-    /core_uploader.sh $S3_BUCKET $S3_KEY_PREFIX
+    export RUN_ID=$(($RANDOM%99999))$(($RANDOM%99999))$(($RANDOM%99999)); \
```
Comment (PettitWesley) [inline scripts/dockerfiles/Dockerfile.init-debug-s3:44]: Yay RUN_ID!

File: troubleshooting/debugging.md [hunk 5]
```diff
@@ -525,11 +525,11 @@ CMD valgrind --leak-check=full --error-limit=no /fluent-bit/bin/fluent-bit -c /f
...
+The best option, which is most likely to catch any leak or segfault is to create a fresh build of the image using the [`Dockerfile.main-debug-valgrind`](https://github.com/aws/aws-for-fluent-bit/blob/mainline/scripts/dockerfiles/Dockerfile.main-debug-valgrind) in AWS for Fluent Bit. This will create a fresh build with debug mode and valgrind support enabled, which gives the highest chance that Valgrind will be able to produce useful diagnostic information about the issue.
 
 1. Check out the git tag for the version that saw the problem
-2. Make sure the `FLB_VERSION` at the top of the `Dockerfile.debug` is set to the same version as the main Dockerfile for that tag. 
-3. Build this dockerfile with the `make debug` target. 
+2. Make sure the `FLB_VERSION` at the top of the `scripts/dockerfiles/Dockerfile.build` is set to the same version as the main Dockerfile for that tag.
+3. Build this dockerfile with the `make debug-valgrind` target.
```
Comment (PettitWesley) [inline troubleshooting/debugging.md]: Can we please briefly note all debug targets here and what they do?

File: troubleshooting/debugging.md [hunk 6]
```diff
@@ -816,11 +816,11 @@ When you recieve a SIGSEGV/crash report from a FireLens customer, perform the fo
 
 ##### 1. Build and distribute a core dump S3 uploader image
 
-You need a customized image build for the specific version/case you are testing. Make sure the `ENV FLB_VERSION` is set to the right version in the `Dockerfile.debug-base` and make sure the `AWS_FLB_CHERRY_PICKS` file has the right contents for the release you are testing. 
+You need a customized image build for the specific version/case you are testing. Make sure the `ENV FLB_VERSION` is set to the right version in the `scripts/dockerfiles/Dockerfile.build` and make sure the `AWS_FLB_CHERRY_PICKS` file has the right contents for the release you are testing. 
 
 Then simply run:
 ```
-make core
+make debug
```
Comment (PettitWesley) [inline troubleshooting/debugging.md:834]: Let's give the names/tags of the outputted images?

File: Makefile (hunk 1 above)
Comment (matthewfala) [reply Makefile:73]: Uh... You mentioned last revision that this is better. After thinking about it for longer, I agree...

Should we bring back a make debug-all?

File: scripts/core_uploader.sh (hunk 2 above)
Comment (matthewfala) [reply scripts/core_uploader.sh:74]: It looks like we're already zipping the file to the crash folder. This line just errors out and does nothing. Checked S3, the .all.zip file is uploaded to s3.

File: scripts/core_uploader.sh (hunk 3 above)
Comment (matthewfala) [reply scripts/core_uploader.sh:63]: This command seems to be a duplicate of the command above that renames the core file to the final name. It's essentially a noop and actually was giving an error message because the file was already renamed. (the rest of the commands still run)

File: troubleshooting/debugging.md (hunk 5 above)
Comment (matthewfala) [reply troubleshooting/debugging.md]: Sure. Added.

File: troubleshooting/debugging.md (hunk 6 above)
Comment (matthewfala) [reply troubleshooting/debugging.md:834]: Sure!

**Comment by matthewfala** [comment]
done

File: scripts/dockerfiles/Dockerfile.main-debug-s3 [hunk 7]
```diff
@@ -36,7 +36,7 @@ CMD echo "AWS for Fluent Bit Container Image Version `cat /AWS_FOR_FLUENT_BIT_VE
     if [ "$S3_KEY_PREFIX" == "issue" ]; then \
         echo "Note: Please set S3_KEY_PREFIX environment variable to a useful identifier - e.g. company name, team name, customer name"; \
     fi; \
-    export RANDOM_ID_VALUE=$(($RANDOM%99999))$(($RANDOM%99999))$(($RANDOM%99999)); \
-    echo "RANDOM_ID is set to $RANDOM_ID_VALUE"; \
-    /fluent-bit/bin/fluent-bit -c /fluent-bit/etc/fluent-bit.conf; \
-    /core_uploader.sh $S3_BUCKET $S3_KEY_PREFIX
+    export RUN_ID=$(($RANDOM%99999))$(($RANDOM%99999))$(($RANDOM%99999)); \
+    echo "RUN_ID is set to $RUN_ID"; \
+    /fluent-bit/bin/fluent-bit -e /fluent-bit/firehose.so -e /fluent-bit/cloudwatch.so -e /fluent-bit/kinesis.so -c /fluent-bit/etc/fluent-bit.conf; \
```
Comment (PettitWesley) [inline scripts/dockerfiles/Dockerfile.main-debug-s3:41]: Thanks for adding the go plugins!!

Let's add them to the debug-fs too??

File: Makefile (hunk 1 above)
Comment (PettitWesley) [reply Makefile:73]: yes

Comment (matthewfala) [reply Makefile:73]: I brought it back.

File: scripts/dockerfiles/Dockerfile.main-debug-s3 (hunk 7 above)
Comment (matthewfala) [reply scripts/dockerfiles/Dockerfile.main-debug-s3:41]: Added

File: troubleshooting/debugging.md [hunk 8]
```diff
@@ -525,11 +525,20 @@ CMD valgrind --leak-check=full --error-limit=no /fluent-bit/bin/fluent-bit -c /f
...
+
+##### Other Options: Other Debug Builds
+Several debug targets can be built via `make <target-name>`
+Here's a list of debug targets and what they do:
+- `debug` and `debug-s3`: these debug targets upload crash symbols, such as a compressed core dump, stack traces, and crashed Fluent Bit executable to an S3 bucket of your choosing. Symbols are also stored in the `/cores` directory. You can access this via a mounted volume. Stack trace is also printed out on crash. More information on how to use the debug image can be found in: [Tutorial: Debugging Fluent Bit with GDB](tutorials/remote-core-dump/README.md).
+- `debug-fs`: this is a light weight debug target that simply outputs the coredump crash file to the `/cores` directory which can be mounted to by a Docker volume. This image does not upload crash symbols to S3, and does not process stack traces.
+- `debug-valgrind`: this is an image that runs Fluent Bit with valrind to catch leaks and segfaults.
+- `init-debug` and `init-debug-s3`: same as `debug` and `debug-s3` images but with the init process
+- `init-debug-fs`: same as `debug-fs` but with the init process.
```
Comment (matthewfala) [inline troubleshooting/debugging.md:541]: Added some image tag descriptions here

**Comment by matthewfala** [comment]
Looks like I didn't push the changes when I commented done previously. Pushed and ready for another review.

File: scripts/dockerfiles/Dockerfile.main-debug-fs [hunk 9]
```diff
@@ -1,4 +1,4 @@
 FROM amazon/aws-for-fluent-bit:main-debug-base
 COPY ./scripts/dockerfiles/Dockerfile.main-debug-base /Dockerfile.2.main-debug-base
 
-CMD /fluent-bit/bin/fluent-bit -c /fluent-bit/etc/fluent-bit.conf; echo "Waiting 2 minutes for EFS core file transfers"; sleep 120
+CMD /fluent-bit/bin/fluent-bit -e /fluent-bit/firehose.so -e /fluent-bit/cloudwatch.so -e /fluent-bit/kinesis.so -c /fluent-bit/etc/fluent-bit.conf; echo "Waiting 2 minutes for EFS core file transfers"; sleep 120
```
Comment (matthewfala) [inline scripts/dockerfiles/Dockerfile.main-debug-fs:4]: added  the go plugins to fs debug

File: troubleshooting/debugging.md [hunk 10]
```diff
@@ -525,11 +525,22 @@ CMD valgrind --leak-check=full --error-limit=no /fluent-bit/bin/fluent-bit -c /f
...
+- `main-debug-all`: builds `debug-s3`, `debug-fs`, and `debug-valgrind` efficiently without rebuilding dependencies.
+- `init-debug-all`: builds `init-debug-s3`, and `init-debug-fs` efficiently without rebuilding dependencies.
```
Comment (matthewfala) [inline troubleshooting/debugging.md:544]: Here's the main-debug-all and init-debug-all descriptions

File: troubleshooting/debugging.md (hunk 6 above)
Comment (matthewfala) [reply troubleshooting/debugging.md:834]: See line 836, below

Comment (matthewfala) [reply troubleshooting/debugging.md:834]: The resulting image will be tagged `amazon/aws-for-fluent-bit:debug` and `amazon/aws-for-fluent-bit:debug-s3`.

File: troubleshooting/debugging.md [hunk 11]
```diff
@@ -525,11 +525,22 @@ CMD valgrind --leak-check=full --error-limit=no /fluent-bit/bin/fluent-bit -c /f
...
 1. Check out the git tag for the version that saw the problem
-2. Make sure the `FLB_VERSION` at the top of the `Dockerfile.debug` is set to the same version as the main Dockerfile for that tag. 
-3. Build this dockerfile with the `make debug` target. 
+2. Make sure the `FLB_VERSION` at the top of the `scripts/dockerfiles/Dockerfile.build` is set to the same version as the main Dockerfile for that tag.
+3. Build this dockerfile with the `make debug-valgrind` target. The image will be tagged with the `amazon/aws-for-fluent-bit:debug-valgrind` tag.
+
+##### Other Options: Other Debug Builds
+Several debug targets can be built via `make <target-name>`
+Here's a list of debug targets and what they do:
+- `debug` and `debug-s3`: these debug targets upload crash symbols, such as a compressed core dump, stack traces, and crashed Fluent Bit executable to an S3 bucket of your choosing. Symbols are also stored in the `/cores` directory. You can access this via a mounted volume. Stack trace is also printed out on crash. More information on how to use the debug image can be found in: [Tutorial: Debugging Fluent Bit with GDB](tutorials/remote-core-dump/README.md).
```
Comment (PettitWesley) [inline troubleshooting/debugging.md:537]: I think only debug-s3 uploads crash symbols...

Comment (PettitWesley) [reply troubleshooting/debugging.md:537]: Just be slightly more clear

Comment (PettitWesley) [reply troubleshooting/debugging.md:537]: oh sorry... these are aliases so they are both the same thing oops

File: troubleshooting/debugging.md [hunk 12]
```diff
@@ -525,11 +525,22 @@ CMD valgrind --leak-check=full --error-limit=no /fluent-bit/bin/fluent-bit -c /f
 
 ##### Option 2: Debug Build (More robust)
 
-The best option, which is most likely to catch any leak or segfault is to create a fresh build of the image using the [`Dockerfile.debug`](https://github.com/aws/aws-for-fluent-bit/blob/mainline/Dockerfile.debug) in AWS for Fluent Bit. This will create a fresh build with debug mode and valgrind support enabled, which gives the highest chance that Valgrind will be able to produce useful diagnostic information about the issue.
+The best option, which is most likely to catch any leak or segfault is to create a fresh build of the image using the [`Dockerfile.main-debug-valgrind`](https://github.com/aws/aws-for-fluent-bit/blob/mainline/scripts/dockerfiles/Dockerfile.main-debug-valgrind) in AWS for Fluent Bit. This will create a fresh build with debug mode and valgrind support enabled, which gives the highest chance that Valgrind will be able to produce useful diagnostic information about the issue.
```
Comment (PettitWesley) [inline troubleshooting/debugging.md]: let's say to build it using the `make debug-valgrind` target??? 

you can't build the dockerfile directly?

I see below you have this... this text here is just slightly misleading.

Comment (PettitWesley) [reply troubleshooting/debugging.md]: this is is a nit

File: troubleshooting/debugging.md [hunk 13]
```diff
@@ -860,6 +872,23 @@ If you choose to deploy this, for the S3 upload on shutdown to work, you must:
 
 Make sure to edit the `Resource` section with your bucket name. 
 
+**RUN_ID Environment Variable**
+The RUN_ID environment variable is automatically set to a random value per each run and can be referenced in your Fluent Bit configuration file. It can be used to customize the log destination for each `aws-for-fluent-bit` execution, and help you to find the log destination when looking at a specific Fluent Bit service log. The following log line will appear before Fluent Bit is executed: `RUN_ID is set to $RUN_ID`.
```
Comment (PettitWesley) [inline troubleshooting/debugging.md:876]: let's be clear RUN_ID is only a feature for debug

Should we make it a feature for non-debug??

**Review by PettitWesley — APPROVED** [review]
minor comments, thanks for updating the guides!!! this is awesome!

File: troubleshooting/debugging.md (hunk 11 above)
Comment (matthewfala) [reply troubleshooting/debugging.md:537]: Yep!

File: troubleshooting/debugging.md (hunk 13 above)
Comment (matthewfala) [reply troubleshooting/debugging.md:876]: Maybe. But we don't have any customer requests for a feature like this. I don't think people would use it. If we did make it a feature for non-debug, we should probably have a name with the AWS_ prefix maybe AWS_FLB_RUN_ID. I don't really think this feature would be widely used.

File: troubleshooting/debugging.md (hunk 12 above)
Comment (matthewfala) [reply troubleshooting/debugging.md]: updated.

**Comment by matthewfala** [comment]
Thanks Wesley! Glad you like it.

File: troubleshooting/debugging.md (hunk 13 above)
Comment (PettitWesley) [reply troubleshooting/debugging.md:876]: 👍 
ok
